### ISSUE TITLE
Faces.getViewIdWithParams()

### ISSUE THREAD

**Issue by NagyGa1** [omnifaces/omnifaces#288]
Would return Faces.getViewId() + Faces.getViewParameterMap():

"/somepath/somepage.xhtml?param1=value1&param2=value2"

, values encoded with URLEncoder.encode(v, StandardCharsets.UTF_8.toString()).

I am willing to do it, just let me know if it is okay.

Use case: generate ?back=returnhere view parameters for a view that might return to different places. (Let me know if there is already a good method for that)

**Comment by NagyGa1** [comment]
Done it outside, not sure if the substring is the right way of doing it:
```java
	@SuppressWarnings("unchecked")
	public static String getViewIdWithParams() {
		final String bookmarkableUrl = Faces.getBookmarkableURL((List) null, true);
		return bookmarkableUrl.substring(Faces.getRequestContextPath().length());
	}
```

**Comment by BalusC** [comment]
It's not exactly clear what you want to do with this this valud. Do you actually want a bookmarkable URL without context path? Why without context path?

Do note that a "view ID" is not necessarily a valid URL. The "view ID" is the bare path to the XHTML file. But when the FaceServlet is mapped on e.g. *.jsf instead of *.xhtml, then the "view ID" won't work as URL.

**Comment by NagyGa1** [comment]
I woud like to return this value from an action method as an outcome with faces-redirect=true appended.

There is a main page backed by a viewscoped bean that does the editing, and there are detail editors that can be navigated, also viewscoped. Then there is a back button on these detail editors. I could program the return path fixed as well of course.

**Comment by BalusC** [comment]
Maybe `Faces#refreshWithQueryString()` is more useful?

**Comment by NagyGa1** [comment]
Sorry I might be ignorant of something, not getting it.

I have two views:
`/path1/view1.xhtml?param1=value1`
there is a button on this one that goes to
`/path2/view2.xhtml?param2=value2&return=/path1/view1.xhtml?param1=value1` (of course this one encoded properly).

Then the action method in `view2` returns `/path1/view1.xhtml?param1=value1&faces-redirect=true`. This is not an URL but an outcome with parameters.

I need the above to come up with the `return=/path1/view1.xhtml?param1=value1` part on the first page.

Maybe this whole concept is bad?

**Comment by BalusC** [comment]
Ahh okay. In effects you thus want something like this?

```
Faces.redirect("path2/view2.xhtml?param2=%s&return=%s", "value2", Faces.getRequestURIWithQueryString());
```

**Comment by NagyGa1** [comment]
For some reason, `Faces.getRequestURIWithQueryString()` = `/baseurl/path1/view1.xhtml`, without the query string part, even thought the URL in the browser is the full thing, not sure if it is a bug.

`Faces.getBookmarkableURL((List) null, true)` returns properly including the parameters, probably because it goes to the Application's viewHandler for the values, while `Faces.getRequestURIWithQueryString()` goes to the servlet.

The action method is called by an ajax commandbutton, PrimeFaces.

I would also need to append the baseUrl to the redirect, and come back from the other view with a redirect as well. Both view's ViewScoped beans will stay in memory due to the redirect, while with action method navigation these would be released.

I could store the query string at the viewaction / prerender of view1, then
```java
return String.format("path2/view2.xhtml?param2=%s&return=%s&%s", "value2", Faces.viewId(), storedQueryString);
````
from the action method.

**Comment by BalusC** [comment]
Indeed, methods which obtain the query string won't work when you aren't using `<o:form includeViewParams>` or `<o:form useRequestURI>`. Sorry, I overlooked that bit.

My previous suggestion should work if you are using `<o:form includeViewParams="true">`.

**Comment by BalusC** [comment]
Added. It's available in today's latest [2.5-SNAPSHOT](https://oss.sonatype.org/content/repositories/snapshots/org/omnifaces/omnifaces/2.5-SNAPSHOT/).

**Comment by NagyGa1** [comment]
Sorry just got around to try it.

Works like charm.
